### PR TITLE
add AAInstrumentation

### PR DESCRIPTION
**Pull request by webmiche**
This is the start of the python instrumentation. This PR adds a baseclass that will be extended with common functionality shared with other specific kinds of instrumentation. Currently, this only implements the generation of the base .bc files for spec.

### CODE REVIEW

**Comment by thetheodor** [comment]
Btw, I completely dropped the `./spec` part of `specbuilder`. Now it's completely done via `ninja`.

I can make it behave more like a library. Then there won't be a need for subprocess etc

**Comment by webmiche** [comment]
Sure, woudl be cool if we could set it up like this. I guess we can just change these parts of the core file when specbuilder has changed.

File: AAInstrumentation/core.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        help="root for execution",
+        default="/home/michel/ETH/AST/llvm-project/",
+    )
+
+    arg_parser.add_argument(
+        "--specbuild_dir",
+        type=str,
+        nargs="?",
+        help="path to specbuilder",
+        default="/home/michel/ETH/AST/specbuilder/",
```
Comment (thetheodor) [inline AAInstrumentation/core.py]: instead of hardcoding such defaults maybe use an argparse file?

File: AAInstrumentation/core.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+    This class is the base class for all instrumentation drivers. It
+    contains the common functionality, like generating the baseline,
+    evaluating results, run a single instrumented compilation, etc. In
+    particular, this contains class variables for all the paths that have to
+    be specified to run the instrumentation on different machines and
+    benchmarks.
+    """
+
+    """
+    The path to the instrumented opt, clang, and llc binaries.
```
Comment (thetheodor) [inline AAInstrumentation/core.py]: Regarding python documentation, it's more idiomatic (and better supported by tools) to add the attribute documentation in the class docstring, e.g.: https://github.com/DeadCodeProductions/diopter/blob/main/diopter/compiler.py#L156

File: AAInstrumentation/core.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+    opt_flag: str
+
+    def generate_baseline(self):
+        run(
+            [
+                "./spec",
+                "build",
+                str(self.initial_dir),
+                str(self.benchmark),
+                str(self.instr_path.joinpath("clang")),
```
Comment (thetheodor) [inline AAInstrumentation/core.py:49]: does this work for c++ files? (i.e., can you link them?)

File: AAInstrumentation/core.py (hunk 1 above)
Comment (webmiche) [reply AAInstrumentation/core.py]: I ended up doing something slightly different: I now pull in the config file to the arg_parse method as well as sys.argv. This means that we can  still run the is `python AAInstrumenation/core.py`, where as before we had to run `python AAInstrumentation/cor.py @config.txt`. Also, if additional arguments are passes in (like `instr_path`), these are overriden.

File: AAInstrumentation/core.py (hunk 3 above)
Comment (webmiche) [reply AAInstrumentation/core.py:49]: I never had in issue with this and I have tried to link before. Given that we will fundamentally change spec builder in the future, would it be fine to delay this issue until after that change?

Comment (webmiche) [reply AAInstrumentation/core.py:49]: I can also open an issue or add a TODO note to the code

**Review by thetheodor — APPROVED** [review]

**Pull request merged**